### PR TITLE
bpo-26219: per opcode cache for LOAD_GLOBAL

### PR DESCRIPTION
**Pull request by methane**
This patch implements opcache only for LOAD_GLOBAL to ease review.

Based on Yury's opcache3.patch.

https://bugs.python.org/issue26219

### CODE REVIEW

**Comment by methane** [comment]
Benchmarks: https://github.com/methane/sandbox/tree/master/2019/opcache_load_global#opcache-for-load_global

File: Lib/test/test_dict_version.py [hunk 1]
```diff
@@ -80,14 +80,14 @@ def test_setitem_same_value(self):
 
         # setting a key to the same value with dict.__setitem__
         # must change the version
-        self.check_version_changed(d, d.__setitem__, 'key', value)
```
Comment (methane) [inline Lib/test/test_dict_version.py:83]: @zware As far as git log, these tests are committed by you.
Do you remember why setting same value should increment the version?

File: Include/code.h [hunk 2]
```diff
@@ -48,6 +50,12 @@ typedef struct {
        Type is a void* to keep the format private in codeobject.c to force
        people to go through the proper APIs. */
     void *co_extra;
+
+    /* Opcodes just-in-time cache */
+    unsigned char *co_opt_opcodemap;
```
Comment (pitrou) [inline Include/code.h]: Can you add comments to explain those field?

File: Include/internal/pycore_code.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#ifndef Py_INTERNAL_CODE_H
+#define Py_INTERNAL_CODE_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+ 
+typedef struct {
```
Comment (pitrou) [inline Include/internal/pycore_code.h:7]: Can you add comments for the fields of those private structs?

File: Objects/codeobject.c [hunk 4]
```diff
@@ -220,9 +222,56 @@ PyCode_New(int argcount, int kwonlyargcount,
...
+    Py_ssize_t opts = 0;
+
+    for (Py_ssize_t i = 0; i < co_size;) {
+        unsigned char opcode = _Py_OPCODE(opcodes[i]);
+        i++;  // 'i' is now aligned to ceval/INSTR_OFFSET()
+
+        // TODO: LOAD_METHOD, LOAD_ATTR
+        if (opcode == LOAD_GLOBAL) {
+            co->co_opt_opcodemap[i] = ++opts;
+            if (opts > 254) {
```
Comment (pitrou) [inline Objects/codeobject.c:265]: Perhaps use 16-bit indices instead?

File: Objects/codeobject.c [hunk 5]
```diff
@@ -220,9 +222,56 @@ PyCode_New(int argcount, int kwonlyargcount,
...
+                break;
+            }
+        }
+    }
+
+    if (opts) {
+        co->co_opt = (_PyOpCodeOpt *)PyMem_Calloc(opts, sizeof(_PyOpCodeOpt));
+        if (co->co_opt == NULL) {
+        PyMem_FREE(co->co_opt_opcodemap);
```
Comment (pitrou) [inline Objects/codeobject.c]: Misindent here.

File: Python/ceval.c [hunk 6]
```diff
@@ -920,6 +967,58 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
+                } \
+            } \
+        } \
+    } while (0)
+
+#define OPCODE_CACHE_DEOPT() \
+    do { \
+        if (co_opt != NULL) { \
+            co_opt->optimized = -1; \
+            co->co_opt_opcodemap[next_instr - first_instr] = 0; \
```
Comment (pitrou) [inline Python/ceval.c]: If you're setting the index to 0 (empty), it means you can't reopt afterwards, since the original index for this opcode is lost. Is that right?

File: Python/ceval.c [hunk 7]
```diff
@@ -95,6 +96,22 @@ static long dxp[256];
 #endif
 #endif
 
+/* opcode cache */
+#define OPCODE_CACHE_MIN_RUNS 1024
```
Comment (pitrou) [inline Python/ceval.c]: Comments?

**Review by pitrou — COMMENTED** [review]
Interesting infrastructure. I've a few comments below.

File: Objects/codeobject.c (hunk 4 above)
Comment (methane) [reply Objects/codeobject.c:265]: To maximize "performance benefit / memory usage", I prefer using 1 byte for each instructions, instead of two bytes.

File: Include/code.h [hunk 8]
```diff
@@ -49,6 +51,14 @@ typedef struct {
        Type is a void* to keep the format private in codeobject.c to force
        people to go through the proper APIs. */
     void *co_extra;
+
+    /* Per opcodes just-in-time cache */
+    // To reduce cache size, we use indirect mapping from opcode index
+    // to cache object: co_opcache[co_opcache_map[next_instr/sizeof(_Py_CODEUNIT)]]
+    unsigned char *co_opcache_map;
+    _PyOpcache *co_opcache;
+    int32_t co_opcache_flag;  // used to determine when create cache
```
Comment (JimJJewett) [inline Include/code.h]: (1) It looks like this is a usage count, rather than a flag.  Maybe call it something like co_run_count?
(2) I don't think you need a full 32 bit int -- which means that it could fit into the same word as the following field.  (and still leave room for some actual flags or something later, even on 32 bit)

File: Include/internal/pycore_ceval.h [hunk 9]
```diff
@@ -28,6 +28,9 @@ PyAPI_FUNC(void) _PyEval_SignalAsyncExc(
 PyAPI_FUNC(void) _PyEval_ReInitThreads(
     _PyRuntimeState *runtime);
 
+/* Private function */
+void _PyEval_Fini(void);
```
Comment (JimJJewett) [inline Include/internal/pycore_ceval.h:35]: If it is really private, why include it in the header, instead of just putting the prototype in the .c file?
Also, why call call it a generic ...Eval_Fini as opposed to ...code_stats ?

File: Include/internal/pycore_code.h [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    PyObject *ptr;  /* Cached pointer (stealed reference) */
```
Comment (JimJJewett) [inline Include/internal/pycore_code.h]: stolen reference?  I'm more used to seeing that around something that can be called ... do you just mean that the object pointed to SHOULD not get an incref for this?  (And is that to avoid circular references?)

File: Lib/test/test_dict_version.py (hunk 1 above)
Comment (JimJJewett) [reply Lib/test/test_dict_version.py:83]: I can think of reasons to prefer it either way ... does the existence of these tests mean there is a backwards compatibility note needed?

File: Objects/codeobject.c [hunk 11]
```diff
@@ -234,9 +236,56 @@ PyCode_New(int argcount, int posonlyargcount, int kwonlyargcount,
     co->co_zombieframe = NULL;
     co->co_weakreflist = NULL;
     co->co_extra = NULL;
+
+    co->co_opcache_flag = 0;
+    co->co_opcache_size = 0;
+    co->co_opcache_map = NULL;
+    co->co_opcache = NULL;
```
Comment (JimJJewett) [inline Objects/codeobject.c:240]: Purely as a style question ... why did you change the order of the fields? (as opposed to initializing in definition order)

File: Include/code.h (hunk 8 above)
Comment (methane) [reply Include/code.h]: Yes -- currently.  I don't think current algorithm to determine when enable cache
is stable, it can be reconsidered several times.  So I don't think we need name represents
concrete usage of the variable for now.

File: Include/internal/pycore_ceval.h (hunk 9 above)
Comment (methane) [reply Include/internal/pycore_ceval.h:35]: Because other private functions are put in "internal" headers too.

File: Include/internal/pycore_code.h (hunk 10 above)
Comment (methane) [reply Include/internal/pycore_code.h]: Ah, I was wrong.  It is borrowed reference, not stolen.

File: Objects/codeobject.c (hunk 11 above)
Comment (methane) [reply Objects/codeobject.c:240]: I didn't changed order in here... I changed only order of struct for compactness and
readability.  I didn't changed here because (a) I just missed it and (b) order in here is
not so important than order in struct.

File: Lib/test/test_dict_version.py (hunk 1 above)
Comment (methane) [reply Lib/test/test_dict_version.py:83]: No.  This test is implemented with `_testcapi`, which is backdoor for testing internals.
No one use it from public stable API.

File: Python/ceval.c (hunk 6 above)
Comment (methane) [reply Python/ceval.c]: Yes, but it is not used for LOAD_GLOBAL cache.  I remove this for now.
